### DISCUSSION TITLE
Notes Sep 7th, 2022

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#104] [Advisory Council Meetings]
# First MapLibre Advisory Council

On Wednesday, Sep 7th, 2022, we had the first MapLibre Advisory Council meeting. The Advisory Council is a forum for exchange between our sponsors and the MapLibre community represented by the Governing Board.

Participation in the MapLibre Sponsorship Program ([PDF](https://maplibre.org/files/Word-Doc-MapLibre-Sponsorship-Program.pdf)) allows companies to get a seat in the Advisory Council. We invited our current sponsors AWS and Meta to yesterday's meeting, and were glad that Microsoft also joined this meeting to get an impression on where the MapLibre Organization stands today.

## Introduction of the Governing Board

Oliver welcomed everybody, introduced the new MapLibre Governing Board to everyone, and gave an overview of the governance of the MapLibre Organization.

## Talks Companies

Meta, AWS, and Microsoft then gave a round of talks to answer the following questions:

* People: What is your organizational structure and who has which role?
* Use-Cases: How do you use or intend to use MapLibre software projects?
* Collaboration: Would you like to collaborate on MapLibre open-source projects in public?
* Priorities: Which technical problems would you like to work on first?

### Meta

Marc Regan represented Meta yesterday.

* Marc is an engineering manager at Meta.
* Working on a strategy for displaying maps on web and mobile apps.
* Main engineer in Marc Regan's team who might be working on MapLibre related stuff is Marc Wilson, a graphics expert.
* Eduardo Neerhut is a Program Manager at Meta working on open-source maps technologies.
* Meta contributes to the open-source community the daylight distribution (a cleaned-up OSM data set), RapiD, and mapwith.ai
* Use-cases: maps are in all possible products at Meta on web, mobile, and in the future maybe on wearable devices. It is possible that MapLibre could be used as a default map renderer across Meta's products in the future.
* Collaboration: Meta wants to upstream everything and stay on the public main branch. Meta's mapping strategy is to use open-source wherever possible and to contribute back.
* Priorities:
  * Binary size - Every kilobyte matters for apps like Facebook and Instagram and MapLibre GL Native would have to be as small as possible for them to be usable in production.
  * Performance - In particular, first load time of a map matters.
  * Power budget - In the future on wearable devices power consumption should be low.
  * OpenGL - Worried that Apple might finally remove support for the already deprecated technology.

### AWS

Haowen You represented AWS yesterday.

* Haowen is an engineering manager leading a team that focuses on open-source mapping technology.
* Amazon Location Services has a range of products that use MapLibre.
* They also collaborate with other organizations within Amazon.
* They work in close collaboration with the open-source office at AWS which is led by David Nalley.
* AWS believes in MapLibre for map rendering.
* They plan to:
  * Build community processes that enable engagement of people who are interested in contributing.
  * Make operational contributions such as cloud resources for continuous integration.
  * Make engineering contributions on the source-code.
  * Help build a community.
* Their interests are:
  * High-priority on Metal shaders implementation on iOS. And possibly Vulkan for Android in the future.
  * Stability and performance improvements in Native.
  * Navigation.
  * Plugin Ecosystem - Work towards a compacted core in Native which is stable, well tested, and easy to maintain. Additional functionality and platform bindings exist around it with for example single-purpose plugins.

### Microsoft

Jesse Levine represented Microsoft yesterday.

* Jesse is a product manager at Microsoft working with geospatial technologies.
* Bing and Azure Maps are one big team working on mapping.
* They cover the needs of consumers and enterprise customers.
* Microsoft contributed data such as the global buildings extract in the past to the open-source community.
* Collaboration:
  * Microsoft wants to be more involved with the MapLibre community.
  * They are committed to MapLibre for web.
  * Microsoft also has mobile ambitions and they are currently evaluating if MapLibre GL Native could serve their needs here.
  * They think that together with the community they might achieve the best engineering result and could potentially benefit from shared maintenance.
* Requirements:
  * Performance is important both on web and native.
  * Microsoft sees a clear need for Metal support on native. Possibly also Vulkan on Android in the future.
  * Are interested in high-detailed maps in the future.
  * Aim for a small, compact core with more features around it.

## Discussion

After the round of talks by the companies we transitioned into a discussion.

Haowen: Is there still telemetry in web or native?
* Everybody: No.

Haowen: How should we handle dependencies on Mapbox software libraries?
* For example the mason from Mapbox build system is used but seems deprecated.
* Birk: Mapbox stopped maintaining many projects we depend on and we can't upstream improvements to these projects.
* Birk: In web we already migrated or removed multiple Mapbox dependencies, and in native the same is happening.

Haowen: What is the general feeling for replacing `mapbox` with `maplibre` in the code bases?
* Harel: In v1 of MapLibre GL JS, mapbox was there but in v2 many occurrences were removed and the goal is to remove them completely.
* Petr: In native the effort was not pushed as far as in web to make initial adoption easier. But the goal should be to move away from mapbox naming and go to maplibre naming.

Haowen: How can we enable contributions in particular in native from companies with private forks?
* Petr: I think this is a cycle. We get contributions with adoption. Adoption comes when there is stability. Sponsorship enables stability, which enables contributions. So everything is coupled and it is great that AWS and Meta made the initial step to bring stability to the MapLibre Organization via their sponsorship!

Petr: If native binary size is a priority - shall we loosen the old paradigm of Mapbox that maps should be pixel-identical between native and web for the same styles?
* Font rendering with SDF is an example - where iOS and Android platform specific fonts may be deployed - as an compile time option 
* This could reduce significantly binary size and simplify development of native.
* It is similar with other cross-platform components
* Remained an open question in yesterday's discussion.

Haowen: VS Code has a great community-driven decision process.
* We could maybe learn from them.
* Example: https://github.com/microsoft/vscode/wiki/development-process

Haowen: Metal
* Is anyone working on a Metal implementation internally already?
  * Not now. A ready to use MetalANGLE branch is at https://github.com/maplibre/maplibre-gl-native/tree/metal-support in case Apple iOS removes OpenGL.
* What are realistic timescales for a true Metal implementation?
  * People did not know.
* Rob: Apple might not have removed support yet for OpenGL, but have the impression that with every iOS release, they break something with OpenGL.
* Nathan: Google maps has a run-time flag to switch between Metal and OpenGL. Should we do the same?
  * Binary size might benefit from a compile-time flag instead.
* Cory: Customers reported that the MetalANGLE, which already is implemented in MapLibre GL Native, is not a robust solution and they would instead prefer a real Metal implementation.
* Haowen: How should people working on a Metal implementation interact with the community?
  * Petr: Like everybody else, i.e., by participating in discussions on GitHub and by attending the Technical Steering Committee meetings.
  * Rob: In the MetalANGLE implementation, Szymon wrote the code and others helped writing documentation and tests.
* Haowen: Worried that the Metal implementation has a large scope and there might be a lack of expertise in the community.
  * Oliver: If experts from other third-parties could help review pull requests of the Metal implementation, this would be a huge benefit for everyone.
* Haowen: It could be a problem if people help review the Metal implementation but, since it will be a long effort, might move on and do other things half-way through. How can we make sure that we have reviewers with the full context of the migration?
  * Luke: One of the highest priorities of the Governing Board is to hire a maintainer for MapLibre GL Native. This person will then be able to follow the entire development process and have the context of previous decisions in mind.

Haowen: How can we measure the success of the MapLibre Organization?
* Luke: The Governing Board has the exact same question which is important when allocating resources on certain tasks. The Board is looking for a metric or procedure for measuring things like adoption and community involvement.

## Feedback

The MapLibre Governing Board is interested in getting some feedback from Meta, AWS, and Microsoft and will send out the following questions to the companies after the meeting:

* What are your thoughts on governance including legal structure, voting, and charter?
* What is your view on the broadening of scope of the project from focus on rendering libraries, to all tools for creating and hosting maps? https://github.com/maplibre/maplibre/discussions/27 which resulted in https://github.com/maplibre/maplibre/blob/main/CHARTER.md#maplibre-organization-goals
* Out of all projects in the MapLibre Organization, what are the higher vs lower priorities in your organization - and therefore your suggested percentage split on focus and resource allocation?
The main areas are: A) GL-JS, B) Native C) Plugins & Extensions D) Research on long-term shared code (code that targets both js & native) E) tile serving tools F) tile generation tools
* What are the MapLibre Organization's strengths, weaknesses, threats, and opportunities?
* Do you have general advices on how we could improve?

### MapLibre Governing Board

* Birk Skyum (MapLibre Governing Board)
* Harel Mazor (MapLibre Governing Board)
* Luke Seelenbinder (MapLibre Governing Board)
* Petr Pridal (MapLibre Governing Board)
* Yuri Astrakhan (MapLibre Governing Board)

### AWS

* Cory Eden (Sr. Engineering Manager at AWS)
* Haowen You (Software Engineering Manager at AWS)
* Seth Fitzsimmons (Principal Engineer at AWS)
* Swagata Prateek (Senior Software Engineer at AWS)

### Meta

* Eduardo Neerhut (Program Manager at Meta)
* Marc Regan (Engineering Manager at Meta)
* Marc Wilson (Software Engineer at Meta)

### Microsoft

* Jesse Levine (Senior Product Manager at Microsoft)
* Marcus Beltman (Software Engineer at Microsoft)
* Taras Vozniuk (Senior Software Engineer at Microsoft)

### MapLibre Technical Steering Committee

* Oliver Wipfli (MapLibre)
* Rob Chohen (MapLibre)
* Matt Logan (Senior Android Engineer at Lyft)
* Nathan Van Fleet (Lead Staff Software Engineer at Lyft)

**Comment by maxammann** [comment]
### Comments

> Is anyone working on a Metal implementation internally already? 

If maplibre-rs counts as an implementation, then technically yes. Metal is supported "first-class". 

> What are realistic timescales for a true Metal implementation? 

It basically requires a complete rewrite -> 2-5 years until it would be production ready.

> Haowen: How should people working on a Metal implementation interact with the community? 

I kind of agree that such topics go beyond discussions and changes through Pull Requests. The  Technical Steering Committee will be the right place to coordinate such changes.

> Haowen: How can we measure the success of the MapLibre Organization?

There is a management tool called OKRs (Objectives and key results). Such frameworks could support us in finding the right goals and metrics and communicating them clearly. It could be difficult to define such metrics as a community though. Also, the governing board should not define all goals top-down. While there are probably goals which can be defined top-down (community involvement) there are others which should be left to the community (metal migration).

> Haowen: VS Code has a great community-driven decision process.

I would like to point out that VSCode is not completely open-source software, as far as I know. At least, the "Microsoft Extension Store" is not usable in non-microsoft builds. I doubt that this decision came from the community.

### General comment

First of all, thank you very much for the great meeting and notes. The information in this post is really valuable! I just have a few concerns regarding transparency with these meetings. I really **do not** want to offend or criticize individuals, instead I want to constructively note a thing.

I would like to raise the question of why Microsoft was invited to the council and nobody from MIERUNE attended. Now, we do not really know from the meeting notes whether MIERUNE was invited. We should definitely add who was invited, additionally to the list to the people who have attended.
I'm asking this question because `Participation in the MapLibre Sponsorship Program` was mentioned just in the sentence before Microsoft is introduced. I feel, that not clearly communicating the reasons for this is unfair to other sponsors.

**Reply by wipfli** [reply]
Thanks @maxammann for the feedback.

The MapLibre Sponsorship Program starts at a yearly contribution of USD 50'000 (Bronze), see https://maplibre.org/files/Word-Doc-MapLibre-Sponsorship-Program.pdf and one of the benefits of participating in the Program is access to the Advisory Council.

I suggest that in the next Advisory Council, which is scheduled for end of November 2022, we only invite current sponsors and the Governing Board. What do you think about this approach, @maxammann?

**Reply by wipfli** [reply]
And sorry if anyone felt excluded... Let's do it better next time!

**Reply by wipfli** [reply]
And to answer your question @maxammann, I invited the people from Microsoft because I think that they can be a valuable party in the MapLibre community and I wanted to help them make an informed decision if becoming part of the MapLibre Sponsorship Program would make sense for them. Now that they have seen how the Advisory Council works, I think they have the full information about the benefits of the Program...

**Reply by maxammann** [reply]
I absolutely understand the rationale. And actually I would welcome inviting **potential** sponsors in the future. For example let's invite potential sponsors like @komoot, or whoever. Maybe such an inclusion could be nice for winning new sponsors.

With the current definition of the council that is not possible. I guess that current sponsors should be fine with that. In fact current sponsors were in the same situation like Microsoft was.
